astropy/nddata/mixins/ndarithmetic.py

```diff
@@ -23,212 +168,638 @@ class NDArithmeticMixin(object):
-                                               mask=operand.mask,
-                                               meta=operand.meta)
-            try:
-                method = getattr(self.uncertainty, propagate_uncertainties)
-                result.uncertainty = method(operand_scaled, result.data)
-            except IncompatibleUncertaintiesException:
-                raise IncompatibleUncertaintiesException(
-                    "Cannot propagate uncertainties of type {0:s} with "
-                    "uncertainties of type {1:s} for {2:s}".format(
-                        self.uncertainty.__class__.__name__,
-                        operand.uncertainty.__class__.__name__,
-                        name))
+            # As with self.uncertainty is None but the other way around.
+            operand.uncertainty = self.uncertainty.__class__(None)
+            result_uncert = self.uncertainty.propagate(operation, operand,
+                                                       result, correlation)
+            operand.uncertainty = None
+            return result_uncert
+
+        else:
+            # Both have uncertainties so just propagate.
+            return self.uncertainty.propagate(operation, operand, result,
+                                              correlation)
+
+    def _arithmetic_mask(self, operand, **kwds):
+        """
+        Calculate the resulting mask
+
+        This is implemented as the piecewise ``or`` operation if both have a
+        mask.
+
+        Parameters
+        ----------
+        operand: `NDData`-like instance
+            The second operand wrapped in an instance of the same class as
+            self.
+
+        Returns
+        -------
+        result_mask: `bool` or `~numpy.ndarray` or None
+            The mask created by a piecewise ``or`` operation. If only one mask
+            was present this mask is returned instead. If neither had a mask
+            ``None`` is returned. This kind of mask combination is only
+            possible if both operands have a mask and this mask is either
+            a boolean or an `~numpy.ndarray` containing booleans.
+        """
 
+        # If only one mask is present we need not bother about any type checks
         if self.mask is None and operand.mask is None:
-            result.mask = None
+            return None
         elif self.mask is None:
-            result.mask = operand.mask.copy()
+            # Make a copy so there is no reference in the result.
+            return deepcopy(operand.mask)
         elif operand.mask is None:
-            result.mask = self.mask.copy()
-        else:  # combine masks as for Numpy masked arrays
-            result.mask = self.mask | operand.mask  # copy implied by operator
+            return deepcopy(self.mask)
+        else:
+            # Make sure these masks are boolean or ndarrays of bools so this
+            # kind of mask operation is possible.
+            if self.mask is not None:
```

Comment:
This again shows that I believe the individual components of `NDData` need to decide if they can be added/subtracted/multiplied/divided. This would only mean that one moves a lot of the logic into light classes or the existing classes. another thing shouldn't it be `and` instead of `or`.